web/packages/quick-editor/src/quick-editor-core.ts

```
@@ -0,0 +1,74 @@
+export type Scope = ( 'about' | 'avatars' | 'verified-accounts' | 'links' | 'interests' | 'contact-info' | 'wallet' | 'photos' | 'design' | 'privacy' )[];
+
+export type ProfileUpdatedType = 'avatar_updated';
+
+export type OnProfileUpdated = ( type: ProfileUpdatedType ) => void;
+
+export type OnOpened = () => void;
+
+export type QuickEditorCoreOptions = Partial<{
+	email: string;
+	scope: Scope;
+	local: string;
+	onProfileUpdated: OnProfileUpdated;
+	onOpened: OnOpened;
+}>;
+
+export class GravatarQuickEditorCore {
+	_name: string;
+	_email: string;
+	_scope: Scope;
+	_local: string;
+	_onProfileUpdated: OnProfileUpdated;
+	_onOpened: OnOpened;
+
+	constructor( { email, scope = [], local, onProfileUpdated, onOpened }: QuickEditorCoreOptions ) {
+		this._name = this._getName();
+		this._email = email;
+		this._scope = scope;
+		this._local = local;
+		this._onProfileUpdated = onProfileUpdated;
+		this._onOpened = onOpened;
+
+		window.addEventListener( 'message', this._onMessage.bind( this ) );
+	}
+
+	open( email?: string ) {
+		email = email || this._email;
+
+		if ( ! email ) {
+			console.error( 'Gravatar Quick Editor: Email not provided' );
+			return;
+		}
+
+		const width = 400;
+		const height = 720;
+		const left = window.screenLeft + ( window.outerWidth - width ) / 2;
+		const top = window.screenTop + ( window.outerHeight - height ) / 2;
+		const options = `popup,width=${ width },height=${ height },top=${ top },left=${ left }`;
+		const host = this._local ? `https://${ this._local }.gravatar.com` : 'https://gravatar.com';
+		const url = `${ host }/profile?email=${ email }&scope=${ this._scope.join( ',' ) }&is_quick_editor=true`;
```

We could check the `this._scope`'s values (and prevent the incorrect values are passed to gravatar.com) for developers, especially for non-ts developers.

---

I'm not sure we need to worry about that. 
An incorrect value would be ignored by the profile editor as it wouldn't match any of the routes.

---

Yes, but JS developers may silently fail to get the correct scope. We can do the enhancement in the next iteration.

---

I might have failed to understand your suggestion.
How would checking the `scope` before opening the popup help the developers guess the correct scopes?

---

Yep, my original thoughts are both:

- (Recommend) Check the incorrect value for developers
- (Good to have) Remove the incorrect values

---

Addressed on [7b1ce7e](https://github.com/Automattic/gravatar/pull/14/commits/7b1ce7e61080364a77dbfd5e99a9f73208409360)